src/freenas/usr/local/libexec/freenas-debug/smart/smart.sh

```diff
@@ -81,21 +81,83 @@ smart_func()
 	ORDER BY +id"
 	section_footer
 
-	section_header "smartctl -a"
-	if [ -f /tmp/smart.out ]; then
-		rm -f /tmp/smart.out
-	fi
+	section_header "smartctl output"
+	rm -f /tmp/smart.out 2>/dev/null
 
 	if is_linux; then
-		disks=$(lsblk -ndo name | grep -v '^sr')
+		# from /proc/devices
+		# only care about sd/vd/nvme major numbers
+		# sd* = 8 - 135
+		# vd* (virtio) = 254
+		# nvme* = 259
+		include="8,65,66,67,68,69,70,71,128,129,130,131,132,133,134,135,254,259"
+		disks=$(lsblk -ndo name -I $include)
 	else
 		disks=$(sysctl -n kern.disks)
 	fi
 
-	for i in $disks
-	do
-    		echo /dev/$i >> /tmp/smart.out
-		smartctl -a /dev/$i >> /tmp/smart.out
+	# SAS to SATA interposers could be involed. Unfortunately,
+	# there is no "easy" way of identifying that there is
+	# one involved without doing some extravagant reading of
+	# specific VPD pages from the device itself. Even doing
+	# that is fraught with errors because the interposer could
+	# mistranslate those pages.
+	# So, instead, we'll try to tell smartctl to do the translation
+	# and if it fails (which it will on proper SAS devices) then
+	# we'll try to run it without translation
+	for i in $disks; do
+		case "$i" in
+		    nvd*)
+			# freebsd returns nvd* devices for the
+			# nvme drives so map to their appropriate
+			# namespace device
+			ns=$(nvmecontrol nsid $i | awk '{print $1}')
+			rc=$?
```

Comment:
william@DESKTOP-OV98RES[~]: foobar | awk '{print $1}'
zsh: command not found: foobar
william@DESKTOP-OV98RES[~]: echo $?
0